Panic reconfiguring v0.11.1 Logging Backend to have 3 persistent replicas instead of 3 ephemeral replicas

The following panic was encountered when I attempted to reconfigure the v0.11.1 Logging Backend to have 3 persistent replicas instead of 3 ephemeral replicas:

```
[04:08:04] INFO Generating certificates {"controller": "opensearchcluster", "controllerGroup": "[opensearch.opster.io](http://opensearch.opster.io/)", "controllerKind": "OpenSearchCluster", "OpenSearchCluster": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "11a160f3-966e-4979-b16f-b5848f1a51a6", "interface": "http"}
[04:08:04] INFO updating existing ism {"controller": "multiclusterrolebinding", "controllerGroup": "[logging.opni.io](http://logging.opni.io/)", "controllerKind": "MulticlusterRoleBinding", "MulticlusterRoleBinding": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "c3431a0e-56c2-4e53-b119-21012136273a", "policy": "log-policy"}
[04:08:04] INFO Observed a panic in reconciler: runtime error: index out of range [0] with length 0 {"controller": "opensearchcluster", "controllerGroup": "[opensearch.opster.io](http://opensearch.opster.io/)", "controllerKind": "OpenSearchCluster", "OpenSearchCluster": {"name":"opni","namespace":"opni"}, "namespace": "opni", "name": "opni", "reconcileID": "11a160f3-966e-4979-b16f-b5848f1a51a6"}
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0
goroutine 880 [running]:
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1())
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115) +0x1e5
panic({0x488cea0?, 0xc003f9b260?})
        runtime/panic.go:914 +0x21f
[opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).reconcileNodeStatefulSet(_](http://opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).reconcileNodeStatefulSet(_), {{0xc001b3818c, 0x4}, 0x3, {0xc001b38190, 0x4}, {0xc0016ffc20, 0xc0016ffc50, {0x0, 0x0, ...}}, ...}, ...)
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/pkg/reconcilers/cluster.go:308 +0x1f18
[opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).Reconcile(0xc00046abd0)](http://opensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).Reconcile(0xc00046abd0))
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/pkg/reconcilers/cluster.go:120 +0x9b0
[opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).reconcilePhaseRunning(0xc001030a50](http://opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).reconcilePhaseRunning(0xc001030a50), {0x63c5490, 0xc0016ff890})
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/controllers/opensearchController.go:320 +0x7ad
[opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).Reconcile(0xc001030a50](http://opensearch.opster.io/controllers.(*OpenSearchClusterReconciler).Reconcile(0xc001030a50), {0x63c5490, 0xc0016ff890}, {{{0xc001b380e4, 0x4}, {0xc001b380e0, 0x4}}})
        opensearch.opster.io@v0.0.0-00010101000000-000000000000/controllers/opensearchController.go:141 +0x779
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x63c5490](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x63c5490)?, {0x63c5490?, 0xc0016ff890?}, {{{0xc001b380e4?, 0x40b50a0?}, {0xc001b380e0?, 0x6399528?}}})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118) +0xb7
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000fafb80](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000fafb80), {0x63c54c8, 0xc000f65860}, {0x45f69c0?, 0xc001b8c020?})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314) +0x365
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000fafb80](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000fafb80), {0x63c54c8, 0xc000f65860})
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265) +0x1c9
[sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2())
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226) +0x79
created by [sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2](http://sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2) in goroutine 375
        [sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222](http://sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222) +0x565
```

--- a/go.mod
+++ b/go.mod
@@ -555,7 +555,7 @@ replace (
 	github.com/rancher/charts-build-scripts => github.com/dbason/charts-build-scripts v0.3.4-0.20220429024555-807c076e8116
 	go.uber.org/zap => github.com/kralicky/zap v1.24.1-0.20230718165024-a2256218e4cc
 	google.golang.org/protobuf => github.com/kralicky/protobuf-go v0.0.0-20230320172414-3920b92ef96f
-	opensearch.opster.io => github.com/dbason/opensearch-k8s-operator/opensearch-operator v0.0.0-20230809020457-b60fd235fc1d
+	opensearch.opster.io => github.com/dbason/opensearch-k8s-operator/opensearch-operator v0.0.0-20230911044558-c677eb01b26a
 )
 
 // Cortex replacements (copied from cortex go.mod)

--- a/plugins/logging/pkg/errors/errors.go
+++ b/plugins/logging/pkg/errors/errors.go
@@ -11,6 +11,7 @@ import (
 var (
 	ErrInvalidList             = utilerrors.New(codes.FailedPrecondition, errors.New("list did not return exactly 1 result"))
 	ErrInvalidPersistence      = utilerrors.New(codes.InvalidArgument, errors.New("invalid persistence config"))
+	ErrInvalidDataPersistence  = utilerrors.New(codes.InvalidArgument, errors.New("minimum of 2 data nodes required if no persistent storage"))
 	ErrClusterIDMissing        = utilerrors.New(codes.InvalidArgument, errors.New("request does not include cluster ID"))
 	ErrOpensearchResponse      = utilerrors.New(codes.Unavailable, errors.New("opensearch request unsuccessful"))
 	ErrNoOpensearchClient      = utilerrors.New(codes.Unavailable, errors.New("opensearch client is not set"))

--- a/plugins/logging/pkg/gateway/admin_v2.go
+++ b/plugins/logging/pkg/gateway/admin_v2.go
@@ -102,7 +102,12 @@ func (m *LoggingManagerV2) CreateOrUpdateOpensearchCluster(ctx context.Context, 
 	}
 	// Input should always have a data nodes field
 	if cluster.GetDataNodes() == nil {
-		return &emptypb.Empty{}, loggingerrors.ErrInvalidDuration
+		return &emptypb.Empty{}, loggingerrors.ErrMissingDataNode
+	}
+
+	// Don't allow a single data node with no persistence
+	if err := m.validateStorage(cluster.GetDataNodes()); err != nil {
+		return nil, err
 	}
 
 	go m.opensearchManager.SetClient(m.managementDriver.NewOpensearchClientForCluster)
@@ -257,6 +262,14 @@ func (m *LoggingManagerV2) validDurationString(duration string) bool {
 		return false
 	}
 	return match
+}
+
+func (m *LoggingManagerV2) validateStorage(dataNodes *loggingadmin.DataDetails) error {
+	if dataNodes.GetReplicas() < 2 && !dataNodes.GetPersistence().GetEnabled() {
+		m.logger.Error("minimum of 2 data nodes required if no persistent storage")
+		return loggingerrors.ErrInvalidDataPersistence
+	}
+	return nil
 }
 
 func (m *LoggingManagerV2) opensearchClusterReady() bool {

